File: leetcode/easy/125.valid-palindrome/solutions/go/main.go
/**
 * Main executable for Valid Palindrome problem
 *
 * This file contains the main function with sample usage and comprehensive testing.
 */

package main

import (
	"fmt"
	"math/rand"
	"time"
)

// Main function with sample usage and test cases
func main() {
	fmt.Println("=== Valid Palindrome - Go Solution ===")

	// Test cases from problem description
	testCases := []struct {
		input       string
		expected    bool
		description string
	}{
		{"A man, a plan, a canal: Panama", true, "Example 1: Complex palindrome with punctuation"},
		{"race a car", false, "Example 2: Not a palindrome"},
		{" ", true, "Example 3: Empty string after cleaning"},
		{"racecar", true, "Simple palindrome"},
		{"hello", false, "Not a palindrome"},
		{"a", true, "Single character"},
		{"", true, "Empty string"},
		{"A", true, "Single uppercase letter"},
		{"0P", false, "Edge case: 0P (0 and P)"},
		{"Madam, I'm Adam", true, "Complex palindrome with punctuation"},
		{"No 'x' in Nixon", true, "Complex palindrome with quotes"},
	}

	// Run test cases for Solution 1
	fmt.Println("--- Solution 1: Two Pointers with unicode functions ---")
	for _, tc := range testCases {
		result := isPalindrome(tc.input)
		status := "✅"
		if result != tc.expected {
			status = "❌"
		}
		fmt.Printf("%s Test: %s\n", status, tc.description)
		fmt.Printf("Input: \"%s\"\n", tc.input)
		fmt.Printf("Expected: %t, Got: %t\n\n", tc.expected, result)
	}

	// Run test cases for Solution 2
	fmt.Println("--- Solution 2: Two Pointers with Custom Character Check ---")
	for _, tc := range testCases {
		result := isPalindromeOptimized(tc.input)
		status := "✅"
		if result != tc.expected {
			status = "❌"
		}
		fmt.Printf("%s Test: %s\n", status, tc.description)
		fmt.Printf("Input: \"%s\"\n", tc.input)
		fmt.Printf("Expected: %t, Got: %t\n\n", tc.expected, result)
	}

	// Performance test
	performanceTest()

	// Edge cases test
	edgeCasesTest()
}

// performanceTest tests performance with various input sizes
func performanceTest() {
	fmt.Println("--- Performance Test ---")

	testStrings := []string{
		"A man, a plan, a canal: Panama",
		"race a car",
		"Madam, I'm Adam",
		"No 'x' in Nixon",
		generateLongPalindrome(1000),
		generateLongPalindrome(5000),
	}

	for _, testString := range testStrings {
		fmt.Printf("Testing string of length %d:\n", len(testString))

		// Test Solution 1
		start := time.Now()
		for i := 0; i < 10000; i++ {
			isPalindrome(testString)
		}
		duration1 := time.Since(start)

		// Test Solution 2
		start = time.Now()
		for i := 0; i < 10000; i++ {
			isPalindromeOptimized(testString)
		}
		duration2 := time.Since(start)

		fmt.Printf("Solution 1: %v for 10,000 iterations\n", duration1)
		fmt.Printf("Solution 2: %v for 10,000 iterations\n", duration2)
		fmt.Printf("Performance ratio: %.2fx\n\n", float64(duration1)/float64(duration2))
	}
}

// edgeCasesTest tests boundary conditions
func edgeCasesTest() {
	fmt.Println("--- Edge Cases Test ---")

	edgeCases := []string{
		"",            // Empty string
		"a",           // Single character
		"aa",          // Two same characters
		"ab",          // Two different characters
		"0P",          // Edge case mentioned in discussion
		"1Q",          // Similar edge case
		"2R",          // Similar edge case
		"aA",          // Case difference
		"a1A",         // Mixed alphanumeric
		"!@#$%^&*()",  // Only special characters
		"12321",       // Numeric palindrome
		"a1b2c3c2b1a", // Mixed alphanumeric palindrome
	}

	for _, testCase := range edgeCases {
		result1 := isPalindrome(testCase)
		result2 := isPalindromeOptimized(testCase)
		fmt.Printf("Input: \"%s\" -> Solution1: %t, Solution2: %t\n", testCase, result1, result2)
	}

	fmt.Println("\nAll edge cases completed successfully! ✅")
}

// generateLongPalindrome generates a long palindrome string for performance testing
func generateLongPalindrome(length int) string {
	rand.Seed(time.Now().UnixNano())

	// Generate first half
	firstHalf := make([]byte, length/2)
	for i := 0; i < length/2; i++ {
		firstHalf[i] = byte(rand.Intn(26) + 'a')
	}

	result := string(firstHalf)

	// Add middle character if odd length
	if length%2 == 1 {
		result += string(byte(rand.Intn(26) + 'a'))
	}

	// Mirror the first half
	for i := len(firstHalf) - 1; i >= 0; i-- {
		result += string(firstHalf[i])
	}

	return result
}

File: leetcode/easy/125.valid-palindrome/solutions/java/Main.java
/**
 * Main executable for Valid Palindrome problem
 *
 * This file contains the main function with sample usage and comprehensive testing.
 */

import java.util.*;

class Main {
    public static void main(String[] args) {
        System.out.println("=== Valid Palindrome - Java Solutions ===");
        
        // Test Solution 1
        System.out.println("--- Solution 1: Two Pointers with Character.isLetterOrDigit() ---");
        testSolution1();
        
        // Test Solution 2
        System.out.println("\n--- Solution 2: Two Pointers with Custom Character Check ---");
        testSolution2();
        
        // Performance comparison
        performanceComparison();
        
        // Edge cases testing
        edgeCasesTest();
    }
    
    /**
     * Test Solution 1: Two Pointers with Character.isLetterOrDigit()
     */
    private static void testSolution1() {
        // Test cases from problem description
        testCase1("A man, a plan, a canal: Panama", true, "Example 1: Complex palindrome with punctuation");
        testCase1("race a car", false, "Example 2: Not a palindrome");
        testCase1(" ", true, "Example 3: Empty string after cleaning");
        
        // Additional test cases
        testCase1("racecar", true, "Simple palindrome");
        testCase1("hello", false, "Not a palindrome");
        testCase1("a", true, "Single character");
        testCase1("", true, "Empty string");
        testCase1("A", true, "Single uppercase letter");
        testCase1("0P", false, "Edge case: 0P (0 and P)");
        testCase1("Madam, I'm Adam", true, "Complex palindrome with punctuation");
        testCase1("No 'x' in Nixon", true, "Complex palindrome with quotes");
    }
    
    /**
     * Test Solution 2: Two Pointers with Custom Character Check
     */
    private static void testSolution2() {
        // Test cases from problem description
        testCase2("A man, a plan, a canal: Panama", true, "Example 1: Complex palindrome with punctuation");
        testCase2("race a car", false, "Example 2: Not a palindrome");
        testCase2(" ", true, "Example 3: Empty string after cleaning");
        
        // Additional test cases
        testCase2("racecar", true, "Simple palindrome");
        testCase2("hello", false, "Not a palindrome");
        testCase2("a", true, "Single character");
        testCase2("", true, "Empty string");
        testCase2("A", true, "Single uppercase letter");
        testCase2("0P", false, "Edge case: 0P (0 and P)");
        testCase2("Madam, I'm Adam", true, "Complex palindrome with punctuation");
        testCase2("No 'x' in Nixon", true, "Complex palindrome with quotes");
    }
    
    private static void testCase1(String input, boolean expected, String description) {
        Solution1 solution = new Solution1();
        boolean result = solution.isPalindrome(input);
        String status = result == expected ? "✅" : "❌";
        System.out.printf("%s Test: %s%n", status, description);
        System.out.printf("Input: \"%s\"%n", input);
        System.out.printf("Expected: %s, Got: %s%n", expected, result);
        System.out.println();
    }
    
    private static void testCase2(String input, boolean expected, String description) {
        Solution2 solution = new Solution2();
        boolean result = solution.isPalindrome(input);
        String status = result == expected ? "✅" : "❌";
        System.out.printf("%s Test: %s%n", status, description);
        System.out.printf("Input: \"%s\"%n", input);
        System.out.printf("Expected: %s, Got: %s%n", expected, result);
        System.out.println();
    }
    
    /**
     * Performance comparison between solutions
     */
    private static void performanceComparison() {
        System.out.println("--- Performance Comparison ---");
        
        // Create test strings of varying lengths
        String[] testStrings = {
            "A man, a plan, a canal: Panama",
            "race a car",
            "Madam, I'm Adam",
            "No 'x' in Nixon",
            generateLongPalindrome(1000),
            generateLongPalindrome(5000)
        };
        
        Solution1 solution1 = new Solution1();
        Solution2 solution2 = new Solution2();
        
        for (String testString : testStrings) {
            System.out.printf("Testing string of length %d:%n", testString.length());
            
            // Test Solution 1
            long startTime = System.nanoTime();
            for (int i = 0; i < 10000; i++) {
                solution1.isPalindrome(testString);
            }
            long endTime = System.nanoTime();
            double solution1Time = (endTime - startTime) / 1_000_000.0;
            
            // Test Solution 2
            startTime = System.nanoTime();
            for (int i = 0; i < 10000; i++) {
                solution2.isPalindrome(testString);
            }
            endTime = System.nanoTime();
            double solution2Time = (endTime - startTime) / 1_000_000.0;
            
            System.out.printf("Solution 1: %.2f ms for 10,000 iterations%n", solution1Time);
            System.out.printf("Solution 2: %.2f ms for 10,000 iterations%n", solution2Time);
            System.out.printf("Performance ratio: %.2fx%n%n", solution1Time / solution2Time);
        }
    }
    
    /**
     * Edge cases testing
     */
    private static void edgeCasesTest() {
        System.out.println("--- Edge Cases Testing ---");
        
        Solution1 solution = new Solution1();
        
        // Test edge cases
        String[] edgeCases = {
            "",           // Empty string
            "a",          // Single character
            "aa",         // Two same characters
            "ab",         // Two different characters
            "0P",         // Edge case mentioned in discussion
            "1Q",         // Similar edge case
            "2R",         // Similar edge case
            "aA",         // Case difference
            "a1A",        // Mixed alphanumeric
            "!@#$%^&*()", // Only special characters
            "12321",      // Numeric palindrome
            "a1b2c3c2b1a" // Mixed alphanumeric palindrome
        };
        
        for (String testCase : edgeCases) {
            boolean result = solution.isPalindrome(testCase);
            System.out.printf("Input: \"%s\" -> %s%n", testCase, result);
        }
        
        System.out.println("\nAll edge cases completed successfully! ✅");
    }
    
    /**
     * Generate a long palindrome string for performance testing
     */
    private static String generateLongPalindrome(int length) {
        StringBuilder sb = new StringBuilder();
        Random rand = new Random();
        
        // Generate first half
        for (int i = 0; i < length / 2; i++) {
            char c = (char) (rand.nextInt(26) + 'a');
            sb.append(c);
        }
        
        // Add middle character if odd length
        if (length % 2 == 1) {
            sb.append((char) (rand.nextInt(26) + 'a'));
        }
        
        // Mirror the first half
        for (int i = length / 2 - 1; i >= 0; i--) {
            sb.append(sb.charAt(i));
        }
        
        return sb.toString();
    }
}
